llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp

```diff
@@ -1739,8 +1738,14 @@ InstructionCost RISCVTTIImpl::getArithmeticInstrCost(
                                                            Op1Info, Op2Info,
                                                            Args, CxtI);
   }
-  return ConstantMatCost +
-         LT.first * getRISCVInstructionCost(Op, LT.second, CostKind);
+
+  InstructionCost InstrCost = getRISCVInstructionCost(Op, LT.second, CostKind);
+  // We use BasicTTIImpl to calculate scalar costs, which assumes floating point
+  // ops are twice as expensive as integer ops. Do the same for vectors so
+  // scalar floating point ops aren't cheaper than their vector equivalents.
+  if (Ty->isFPOrFPVectorTy())
```

Comment:
Mostly stylistic, but I think we should sink this inside getRISCVInstructionCost (probably the default case in the switch), and key this off the legalized type (i.e. LT.second).  I don't think this would currently cause any behavior difference, but we could reasonable add another caller of getRISCVInstructionCost which wants this behavior in the future.

Reply:
That's what I did at first, but after grepping I found other non-arithmetic FP users of getRISCVInstructionCost like VFMV_V_F which had some observable cost model differences.

BasicTTIImpl only seems to double the cost for floats in getArithmeticInstrCost, so I wanted to keep it local to RISCVTTIImpl::getArithmeticInstrCost since that's the only place where we need the fp-cost-doubling to stay in sync with scalar costs

Reply:
Fair enough.

Reply:
Sorry to be late to see this.
Can we do this to specific FP operations like VFADD_VV, VFMUL_VV, VFDIV_VV instead?
I am trying to introduce a vendor-specific TTI cost in #90951. With this patch getArithmeticInstrCost would double the vendor cost  unexpectedly.

Reply:
Would the scalar vendor cost not also be doubled too? Or are there custom vendor FP ops that don't go through BasicTTIImpl that avoid it?

Reply:
So far I am only tuning the vector cost, so I don't have vendor cost for scalar ops.

Reply:
I might be missing something, if that's the case then shouldn't we also be doubling the vendor costs in getArithmeticInstrCost anyway? Otherwise they'll have the same issue before this patch where the vector ops are cheaper than scalars.

Reply:
But putting the doubling logic in getRISCVInstructionCost means we would end up doubling the cost everywhere, I think we only need to double it in RISCVTTIImpl::getArithmeticInstrCost where we invoke BasicTTIImpl:: getArithmeticInstrCost?

Reply:
I don't know if there will be VFADD/VFMUL used in other TTI cost functions.
But we will need to add a check for the cost kind since the logic only applies to TTI::TCK_RecipThroughput but not to TTI::TCK_CodeSize.